doc/design/ops/sequence_decoder.md

```diff
@@ -0,0 +1,345 @@
+# A LoD based Sequence Decoder (Beam Search)
+Sequence Decoder is an operator that help generate sequences, 
+it shares much logic with `dynamic_recurrent_op`,
+but some more codes to support beam search algorithm.
+
+In text generation tasks, such as machine translation, 
+a neural network is trained to rate candidate words given the context,
+a decoder will be used to select out good candidates as the next word and extend the prefix with the candidates.
+
+## Beam Search
+Beam search is a heuristic search algorithm that explores a graph by expanding the most promising node in a limited set.
+
+It is the core component of Sequence Decoder.
+
+In the original implementation of `RecurrentGradientMachine`, the beam search is a method in RNN,
```

Comment:
Are there more than one implementations of `RecrurentGradientMachine`? What is the other implementations besides the "original" one?